File: app/src/main/java/com/orgzly/android/git/GitSSHKeyTransportSetter.java
package com.orgzly.android.git;

import android.net.Uri;
import android.os.Build;

import androidx.annotation.RequiresApi;

import com.orgzly.android.App;

import org.eclipse.jgit.annotations.NonNull;
import org.eclipse.jgit.api.TransportCommand;
import org.eclipse.jgit.api.TransportConfigCallback;
import org.eclipse.jgit.internal.transport.sshd.OpenSshServerKeyDatabase;
import org.eclipse.jgit.transport.SshSessionFactory;
import org.eclipse.jgit.transport.SshTransport;
import org.eclipse.jgit.transport.sshd.ServerKeyDatabase;
import org.eclipse.jgit.transport.sshd.SshdSessionFactory;

import java.io.File;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Collections;
import java.util.List;

public class GitSSHKeyTransportSetter implements GitTransportSetter {
    private final TransportConfigCallback configCallback;

    public GitSSHKeyTransportSetter(String pathToSSHKey) {

        SshSessionFactory factory = new SshdSessionFactory(null, null) {

            @Override
            public File getHomeDirectory() {
                return App.getAppContext().getFilesDir();
            }

            @RequiresApi(api = Build.VERSION_CODES.O)
            @Override
            protected List<Path> getDefaultIdentities(File sshDir) {
                return Collections.singletonList(Paths.get(Uri.decode(pathToSSHKey)));
            }

            @Override
            protected String getDefaultPreferredAuthentications() {
                return "publickey";
            }

            @Override
            protected ServerKeyDatabase createServerKeyDatabase(@NonNull File homeDir,
                                                                @NonNull File sshDir) {
                // We override this method because we want to set "askAboutNewFile" to False.
                return new OpenSshServerKeyDatabase(false,
                        getDefaultKnownHostsFiles(sshDir));
            }
        };

        SshSessionFactory.setInstance(factory);

        // org.apache.sshd.common.config.keys.IdentityUtils freaks out if user.home is not set
        System.setProperty("user.home", App.getAppContext().getFilesDir().toString());

        configCallback = transport -> {
            SshTransport sshTransport = (SshTransport) transport;
            sshTransport.setSshSessionFactory(factory);

        };
    }

    public TransportCommand setTransport(TransportCommand tc) {
        tc.setTransportConfigCallback(configCallback);
        tc.setCredentialsProvider(new SshCredentialsProvider());
        return tc;
    }
}

File: app/src/main/java/com/orgzly/android/git/GitSshKeyTransportSetter.kt
package com.orgzly.android.git

import android.content.Intent
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.orgzly.R
import com.orgzly.android.App
import com.orgzly.android.ui.SshKeygenActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import org.eclipse.jgit.annotations.NonNull
import org.eclipse.jgit.api.TransportCommand
import org.eclipse.jgit.api.TransportConfigCallback
import org.eclipse.jgit.internal.transport.sshd.OpenSshServerKeyDatabase
import org.eclipse.jgit.transport.SshSessionFactory
import org.eclipse.jgit.transport.SshTransport
import org.eclipse.jgit.transport.Transport
import org.eclipse.jgit.transport.sshd.ServerKeyDatabase
import org.eclipse.jgit.transport.sshd.SshdSessionFactory
import java.io.File
import java.security.KeyPair

class GitSshKeyTransportSetter: GitTransportSetter {
    private val configCallback: TransportConfigCallback
    private val activity = App.getCurrentActivity()
    private val context = App.getAppContext()

    init {
        val factory: SshSessionFactory = object : SshdSessionFactory(null, null) {

            override fun getHomeDirectory(): File { return context.filesDir }

            override fun getDefaultPreferredAuthentications(): String { return "publickey" }

            override fun createServerKeyDatabase(
                @NonNull homeDir: File,
                @NonNull sshDir: File
            ): ServerKeyDatabase {
                // We override this method because we want to set "askAboutNewFile" to False.
                return OpenSshServerKeyDatabase(
                    false,
                    getDefaultKnownHostsFiles(sshDir)
                )
            }

            override fun getDefaultKeys(@NonNull sshDir: File): Iterable<KeyPair>? {
                return if (SshKey.exists) {
                    listOf(SshKey.getKeyPair())
                } else {
                    onMissingSshKeyFile()
                    null
                }
            }
        }

        SshSessionFactory.setInstance(factory)

        // org.apache.sshd.common.config.keys.IdentityUtils freaks out if user.home is not set
        System.setProperty("user.home", context.filesDir.toString())

        configCallback = TransportConfigCallback { transport: Transport ->
            val sshTransport = transport as SshTransport
            sshTransport.sshSessionFactory = factory
        }
    }

    override fun setTransport(tc: TransportCommand<*, *>): TransportCommand<*, *> {
        tc.setTransportConfigCallback(configCallback)
        tc.setCredentialsProvider(SshCredentialsProvider())
        return tc
    }

    private fun onMissingSshKeyFile() {
        if (activity != null) {
            val builder = MaterialAlertDialogBuilder(activity)
                .setMessage(R.string.git_ssh_on_missing_key_dialog_text)
                .setTitle(R.string.git_ssh_on_missing_key_dialog_title)
            builder.setPositiveButton(activity.getString(R.string.yes)) { _, _ ->
                val intent =
                    Intent(activity.applicationContext, SshKeygenActivity::class.java)
                activity.startActivity(intent)
            }
            builder.setNegativeButton(activity.getString(R.string.not_now)) {
                    dialog, _ -> dialog.dismiss()
            }
            runBlocking(Dispatchers.Main) { activity.alertDialog = builder.show() }
        }
    }
}